File: src/main/scala/edu/hitsz/aircraft/AbstractAircraft.scala
package edu.hitsz.aircraft

import edu.hitsz.basic.FlyingObject
import edu.hitsz.bullet.AbstractBullet

/**
 * 所有种类飞机的抽象父类：
 * 敌机（BOSS, ELITE, MOB），英雄飞机
 *
 * @author hitsz
 */
abstract class AbstractAircraft(locationX: Int, locationY: Int, speedX: Int, speedY: Int, hpInit: Int)
  extends FlyingObject(locationX, locationY, speedX, speedY) {
  // 生命值
  var hp = hpInit

  def decreaseHp(decrease: Int) = {
    hp -= decrease
    if (hp <= 0) {
      hp = 0
      vanish()
    }
  }

  def getHp = hp

  /**
   * 飞机射击方法，可射击对象必须实现
   *
   * @return
   * 可射击对象需实现，返回子弹
   * 非可射击对象空实现，返回空 Seq
   */
  def shoot(): Seq[AbstractBullet]
}

File: base/src/main/java/edu/hitsz/aircraft/AbstractAircraft.java
package edu.hitsz.aircraft;

import edu.hitsz.basic.AbstractFlyingObject;
import edu.hitsz.bullet.BaseBullet;
import edu.hitsz.prop.AbstractProp;

import java.util.LinkedList;

/**
 * 所有种类飞机的抽象父类：
 * 敌机（BOSS, ELITE, MOB），英雄飞机
 *
 * @author hitsz
 */
public abstract class AbstractAircraft extends AbstractFlyingObject {
    /**
     * 生命值
     */
    protected int maxHp;
    protected int hp;
    protected int score;

    public int getScore() {
        return score;
    }

    public AbstractAircraft(int locationX, int locationY, int speedX, int speedY, int hp, int score) {
        super(locationX, locationY, speedX, speedY);
        this.hp = hp;
        this.maxHp = hp;
        this.score = score;
    }

    public void decreaseHp(int decrease) {
        hp -= decrease;
        if (hp <= 0) {
            hp = 0;
            vanish();
        }
    }

    public int getHp() {
        return hp;
    }


    /**
     * 飞机射击方法，可射击对象必须实现
     *
     * @return 可射击对象需实现，返回子弹
     * 非可射击对象空实现，返回null
     */
    public abstract LinkedList<BaseBullet> shoot();

    /**
     * 飞机掉落道具
     * @return 返回所有的道具
     */
    public abstract LinkedList<AbstractProp> dropProps();
}


